docs/extensibility/extend-rest-api-add-data.md

```diff
@@ -0,0 +1,289 @@
+That data is namespaced to your plugin and protected from other plugins malfunctioning.
+The data is available on all frontend filters and slotFills for you to consume.
+
+## Basic usage
+You can use ExtendRestAPI by registering a couple of functions, `schema_callback` and `data_callback` on a specific endpoint namespace. Those functions are going to be called at execution time and are going to be passed relevant data.
+
+```PHP
+
+use Automattic\WooCommerce\Blocks\Package;
+use Automattic\WooCommerce\Blocks\Domain\Services\ExtendRestApi;
+use Automattic\WooCommerce\Blocks\StoreApi\Schemas\CartSchema;
+
+add_action('woocommerce_blocks_loaded', function() {
+ // ExtendRestApi relies on a Singleton that is shared between the API and consumers.
+ // You shouldn't initiate your own ExtendRestApi using new ExtendRestApi but should always get it from Package dependecy injection manager.
+ $extend = Package::container()->get( ExtendRestApi::class );
+
+ $extend->register_endpoint_data(
+ array(
+ 'endpoint' => CartSchema::IDENTIFIER,
+ 'namespace' => 'plugin_namespace',
+ 'data_callback' => 'my_data_callback',
+ 'schema_callback' => 'my_schema_callback',
+ 'schema_type' => ARRAY_A,
+ )
+ );
+});
+
+
+function my_data_callback() {
+ return [
+ 'custom-key' => 'custom-value';
+ ]
+}
+
+function my_schema_callback() {
+ return [
+ 'custom-key' => [
+ 'description' => __( 'My custom data', 'plugin-namespace' ),
+ 'type' => 'string'
+ 'readonly' => true,
+ ]
+ ]
+}
+```
+
+Data callback and Schema callback can also receive parameters:
+
+```PHP
+
+function my_cart_item_callback( $cart_item ) {
+ $product = $cart_item['data'];
+ if ( is_my_custom_product_type( $product ) ) {
+ $custom_value = get_custom_value( $product );
+ return [
+ 'custom-key' => $custom_value;
+ ]
+ }
+}
+
+```
+
+## API Definition
+
+- `ExtendRestAPI::register_endpoint_data`: Used to register data to a custom endpoint. It takes an array of arguments:
+
+| Attribute | Type | Required | Description |
+| :-------- | :----- | :------: | :------------------------------------ |
+| `endpoint` | string | Yes | The endpoint you're trying to extend, it is suggested that you use `::IDENTIFIER` Available on the route Schema class to avoid typos. |
+| `namespace` | string | Yes | Your plugin namespace, the data will be available under this namespace in the StoreAPI response. |
+| `data_callback` | callback | Yes | A callback that returns an array with your data. |
+| `schema_callback` | callback | Yes | A callback that returns the shape of your data. |
+| `data_type` | string | No (default: `ARRAY_A` ) | The type of your data, if you're adding an object (key => values), it should be `ARRAY_A`, you're adding a list of items, it should be `ARRAY_N`. |
+
+## Putting it all together
+
+This is a complete example that shows how you can register contextual WooCommerce Subscriptions data in each cart item (simplified).
+
+This example uses Formatters, another property documented here (TBD).
```

Comment:
Created follow up issue for Formatters https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4210

Reply:
I'd suggest maybe adding a note to that issue that the information should be cross-linked here once it's in place.